### ISSUE TITLE
In which part of the code that i can find "page" concept?

### ISSUE THREAD

**Issue by YEOMJINSEOP** [mit-han-lab/Quest#18]
Hi, @Sakits, @SiriusNEO, @happierpig.
Thanks for sharing wonderful works.

1. I have a question regarding the "paging" concept in the code.

While i was analyzing codes in quest_attention.py,
i couldn't find any "page" concept, while i could found picking top-k "chunked kv_seq_len" in `def local_heavy_hitter_mask`.
Is chunking concpet in the code is equal to paging in the paper? if not, where can i find "page" concept in the code?

2. in which part of the code uses QuestAttention class? 
i see that LlamaAttention module's forward is substitued to "quest_attention.py's def forward"
However, i don't see any part that load QuestAttention class and use in quest_attention.py. 
Is QuestionAttention class not used in the entire code? 

Thanks!

**Comment by Sakits** [comment]
Hi @YEOMJINSEOP , 

Thank you for your interest in our work!

1. Yes, the “chunk” concept in the code is equivalent to the “page” concept in the paper. The chunking mechanism in `local_heavy_hitter_mask` is the same idea as paging.
2. The `QuestAttention` class is implemented to work with the kernel to enable real acceleration. However, for simplicity during accuracy evaluations, we simulate sparse attention using the implementation in `evaluation/quest_attention.py` instead of invoking the CUDA kernel directly.

I hope this answers your questions! Please let us know if you have any further questions. :)

**Comment by YEOMJINSEOP** [comment]
Hi, @Sakits,

Thank you for your kind and detailed explanations.

I have one more question:
If I want to measure the memory consumption of QUEST and compare it with a naive attention mechanism, is it okay to use the evaluation/quest_attention.py implementation instead of directly invoking the CUDA kernel?

Would this approach yield memory usage results comparable to the real implementation, or do I need to modify the code to incorporate the CUDA kernel for an accurate comparison?

Thank you again for your assistance!

**Comment by happierpig** [comment]
Hi @YEOMJINSEOP ,

Thanks for your questions. Here are my quick answers:
1. Quest will not reduce **_memory capacity usage_**. Instead, Quest boosts inference by reducing memory loading which saves bandwidth resources.
2. To measure the memory loading reduction, using `evaluation/quest_attention.py` will not yield comparable results as it basically computes all results and uses `attention_mask` to ignore some tokens. Only real implementation can showcase the memory loading benefit.

**Comment by YEOMJINSEOP** [comment]
Thanks to @Sakits  and @happierpig for answers! 

It helped me to understand Quest implementation more clearly.

However, one thing i still don't get clearly is "How `evaluation/quest_attention.py` can be used to show Quest's results",
because `evaluation/quest_attention.py` seems not following the algorithm of Quest shown in Figure 5 of paper.
- In the Figure 5 of the paper, at stage 1, Quest have a process "Element-wise Product -> Per-channel Max -> Sum -> Determine Top-K pages."
- On the other hand, in `evaluation/quest_attention.py`, seems to have no "Sum -> Determine Top-K pages" process, but  having  a process of "Get each page's per-channel Max (`max_key`) -> calculate attention scores between` positive_query` and `max_key` -> using Top-K pages having highest attention scores".

Could you explain a bit more about the part I'm confused about and how to use real `QuestAttention ` class for evaluation of PG-19, Pass Key Retrieval and LongBench?

Thanks! : )

**Comment by Sakits** [comment]
Hi @YEOMJINSEOP , 

Thank you for your question! 

The implementation in `evaluation/quest_attention.py` is indeed aligned with the algorithm described in the paper. (i.e. The process `"Element-wise Product -> Per-channel Max -> Sum"` in the paper is equivalent to calculating attention scores between the `positive_query` and the `max_key`. )

Here's how:

Firstly, the `"Element-wise Product" and "Sum"` steps are actually the same calculating the attention score between the query and the key, which is a common operation in attention mechanisms.

For the `"Per-channel Max"`, the purpose is to determine the key that will produce the maximal value in each dimension when multiplied by the query. 

To achieve this:
1. Transform the query into all-positive values by multiplying each of its elements by their sign.
2. To maintain mathematical equivalence of q*k, multiply the keys by the inverse of the query’s sign (which is the same as the query’s sign itself).
3. After this transformation, we can compute the per-channel max directly on the adjusted keys (referred to as max_key) without needing the query.

Therefore, this implementation is **mathematically equivalent but more straightforward** to the process described in the paper. The key steps are:

1. Get the positive values of queries (`positive_query`) and per-channel max for the adjusted keys (`max_key`).
2. Calculate the attention score between the `positive_query` and `max_key` (this corresponds to `"Element-wise Product -> Per-channel Max -> Sum"`).
3. Use these scores to determine the top-K pages.

If you're interested in re-implementing the exact process as described in the paper, you’ll find that the results will exactly match the current implementation in `evaluation/quest_attention.py`. :)

The `QuestAttention` can indeed be used for evaluation. However, the evaluation process using the full `QuestAttention` is more complex, which is why we did not provide scripts for it.

I hope this helps clarify your questions! Let me know if you need further details or have additional concerns.

**Comment by YEOMJINSEOP** [comment]
Hi @Sakits,

Thank you so much for your detailed and kind explanation.
After thoroughly reviewing your response and delving deeper into the code, I now have a clear understanding of how quest_attention operates and how it aligns with the Quest mechanism described in the paper.

Would it be possible to provide an evaluation script using the QuestAttention class? I’m very curious to explore more about how the kernel works during evaluation in detail.

Thank you again, and wishing you a happy end of the year! : )

**Comment by happierpig** [comment]
Hi @YEOMJINSEOP ,

Please take a look at https://github.com/mit-han-lab/Quest/blob/54bbfb00939ef4af9daaa48a81dc7c592e44acf8/scripts/example_textgen.py#L22 which utilizes the kernels for inference. Arbitrary benchmark scripts can be written based on this script. Thanks!
